pelita/ui/qt/qt_viewer.py

```diff
@@ -0,0 +1,316 @@
+import cmath
+import json
+import logging
+import math
+import os
+from pathlib import Path
+import shutil
+import signal
+import sys
+
+import zmq
+
+from PyQt5 import QtCore, QtGui, QtWidgets
+from PyQt5.QtWidgets import QWidget, QApplication, QMainWindow
+from PyQt5.QtCore import QPointF, QRectF
+
+from ...graph import diff_pos
+from ...datamodel import CTFUniverse
+
+_logger = logging.getLogger(__name__)
+_logger.setLevel(logging.DEBUG)
+
+signal.signal(signal.SIGINT, signal.SIG_DFL)
+
+class ZMQListener(QtCore.QThread):
+    message = QtCore.pyqtSignal(str)
+
+    def __init__(self, address, exit_address):
+        super().__init__()
+        self.address = address
+        self.exit_address = exit_address
+        self.running = True
+    
+    def run(self):
+        self.context = zmq.Context()
+        self.socket = self.context.socket(zmq.SUB)
+        self.socket.setsockopt_unicode(zmq.SUBSCRIBE, "")
+        self.socket.connect(self.address)
+
+        self.exit_socket = self.context.socket(zmq.PAIR)
+        self.exit_socket.connect(self.exit_address)
+
+        self.poll = zmq.Poller()
+        self.poll.register(self.socket, zmq.POLLIN)
+        self.poll.register(self.exit_socket, zmq.POLLIN)
+
+        while self.running:
+            evts = dict(self.poll.poll(1000))
+            if self.socket in evts:
+                message = self.socket.recv_unicode()
+                self.message.emit(message)
+            if self.exit_socket in evts:
+                self.running = False
+ 
+
+class Ui_MainWindow(object):
+    def setupUi(self, MainWindow):
+        MainWindow.resize(600, 300)
+        self.centralwidget = QWidget(MainWindow)
+        self.horizontalLayout = QtWidgets.QHBoxLayout(self.centralwidget)
+        MainWindow.setCentralWidget(self.centralwidget)
+        menubar = QtWidgets.QMenuBar(None)
+#        self.menubar.setGeometry(QtCore.QRect(0, 0, 500, 22))
+        fileMenu = menubar.addMenu('&File')
+        
+        impAct = QtWidgets.QAction('Import mail', MainWindow)
+        fileMenu.addAction(impAct)
+        
+        menubar.addMenu(fileMenu)
+
+        MainWindow.setMenuBar(menubar)
+#        self.statusbar = QtWidgets.QStatusBar(MainWindow)
+#        MainWindow.setStatusBar(self.statusbar)
+        QtCore.QMetaObject.connectSlotsByName(MainWindow)
+
+
+class QtViewer(QMainWindow):
+    def __init__(self, address, controller_address=None,
+                       geometry=None, delay=None, export=None,
+                       *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        if export:
+            png_export_path = Path(export)
+            if not png_export_path.is_dir():
+                raise RuntimeError("Not a directory: {png_export_path}")
```

Comment:
shouldn't this get formatted?

Reply:
It’s a commit named ‘WIP’. This will be rebased eventually. Sorry for the noise. ;)